pkg/kubelet/pod_workers.go

```diff
@@ -734,10 +751,75 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
 	return gracePeriod, status.gracePeriod != 0 && status.gracePeriod != gracePeriod
 }
 
+// allowPodStart tries to start the pod and returns true if allowed, otherwise
+// it requeues the pod and returns false.
+func (p *podWorkers) allowPodStart(pod *v1.Pod) bool {
+	if !kubetypes.IsStaticPod(pod) {
+		// TODO: Do we want to allow non-static pods with the same full name?
+		// Note that it may disable the force deletion of pods.
+		return true
+	}
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+	if !p.isStaticPodAllowedToStart(pod) {
+		p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
+		if status := p.podSyncStatuses[pod.UID]; status != nil {
+			status.working = false
+		}
+		return false
+	}
+	p.allowNextStaticPod(kubecontainer.GetPodFullName(pod))
+	return true
+}
+
+// isStaticPodAllowedToStart returns true if there is no other started static
+// pod with the same full name and pod is the first static pod waiting to
+// start.
+func (p *podWorkers) isStaticPodAllowedToStart(pod *v1.Pod) bool {
+	fullname := kubecontainer.GetPodFullName(pod)
+	uid, started := p.startedStaticPodsByFullname[fullname]
+	if started {
+		if uid == pod.UID {
+			// pod is already started
+			return true
+		}
+		return false
+	}
+
+	waitingPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingPods) == 0 {
+		return true
+	}
+
+	return pod.UID == waitingPods[0]
+}
+
+// allowNextStaticPod allows next static pod to start for the given fullname
+func (p *podWorkers) allowNextStaticPod(fullname string) {
+	waitingPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingPods) == 0 {
+		return
+	}
+
+	p.startedStaticPodsByFullname[fullname] = waitingPods[0]
```

Comment:
This seems like it is in the wrong place - this method already knows what the started pod UID is, so the method should be taking `(fullName, uid)` just like `isStaticPodAllowedToStart` and you should be setting this to

```
p.startedStaticPodsByFullName[fullName] = uid
```